docs: MCP clients guides (#56828)
gravitational/teleport: docs/pages/connect-your-client/model-context-protocol/database-access.mdx

---
title: Database Access for MCP
description: How to configure MCP clients to use Teleport databases as MCP servers.
---

> This guide explains how to expose your **PostgreSQL** Teleport databases as MCP servers.

Reviewer: This is a clients guide, right?
```suggestion
This guide explains how to connect to your **PostgreSQL** Teleport databases with MCP clients.
```

### Prerequisites

(!docs/pages/includes/edition-prereqs-tabs.mdx!)

- Teleport Database Service with a PostgreSQL database enrolled. See our [guides](../../enroll-resources/database-access/database-access.mdx)
  for options on how to set up Database Access for PostgreSQL databases, such as
  the [AWS RDS PostgreSQL](../../enroll-resources/database-access/enroll-aws-databases/rds.mdx)
  and [self-hosted PostgreSQL](../../enroll-resources/database-access/enroll-self-hosted-databases/postgres-self-hosted.mdx) guides.

### Security considerations

Since language models can execute any query on your database, we advise creating
a database user with only the permissions you want the models to have. Setting
up a user with read-only permissions will help prevent accidental changes to

> your database.

Author: Updated it and added an example of a PostgreSQL script to create a read-only user. I also added references to auto-user provisioning and DAC.

### Step 1/2. Installation

First, sign in into your Teleport cluster using `tsh login`:

```code
$ tsh login --proxy=<Var name="teleport.example.com:443" /> --user=myuser@example.com
```

Only databases that you have access can be exposed as MCP servers. You can
retrieve the list of databases and their connection options using `tsh db ls`:
  
```code
$ tsh db ls
# Name            Description          Allowed Users Labels 
# --------------- -------------------- ------------- -------
# postgres-dev    Development database [* postgres]  env=dev
# postgres-prod   Production database  [readonly]    env=prod
```

> Each database you want accessible through MCP must be configured separately.

Reviewer: ```suggestion
Each database you want accessible through MCP must be configured with your MCP client individually.
```

This is done using the `tsh mcp db config` command. Like `tsh db connect`, this
command requires you to select the database user and database name that the MCP
connections will use.

This command can either generate a configuration file (using the `mcpServers`
format) for manual MCP client updates or automatically update the MCP client
configuration.

<Tabs>
<TabItem label="Claude Desktop">
`tsh` can automatically update the Claude Desktop MCP configuration file to
include Teleport's configuration:

```code
$ tsh mcp db config --db-user=postgres --db-name=employees --client-config=claude postgres-dev
Added database "postgres-dev" on the client configuration at:

> /path/to/Claude/claude_desktop_config.json

Reviewer: What's the default here?

Author: It depends on the platform. We can default the example to macOS.
